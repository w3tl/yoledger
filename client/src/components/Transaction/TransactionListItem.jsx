import React from 'react';
// import PropTypes from 'prop-types';
import moment from 'moment';
import { transactionPropType } from './propTypes';
import { Button } from '../Elements';
import TransactionForm from './TransactionFormHOC';

export function Item({ // eslint-disable-next-line react/prop-types
  source, destination, amount, date, onClick,
}) {
  return (
    <li style={{ display: 'flex' }}>
      <p><b>From:</b> {source.name}</p>
      <p><b>; To:</b> {destination.name}</p>
      <p><b>; Amount:</b> {amount} </p>
      <p><b>; {moment(new Date(date)).format('DD MMM YY')}</b></p>
      <Button onClick={onClick}>Edit</Button>
    </li>
  );
}

class TransactionListItem extends React.Component {
  static propTypes = {
    transaction: transactionPropType.isRequired,
  }

  state = {
    isEdit: false,
  }

  handleClick = () => {
    this.setState(prevState => ({ isEdit: !prevState.isEdit }));
  }

  render() {
    const { transaction, ...other } = this.props;
    const { isEdit } = this.state;

    if (isEdit) {
      return <TransactionForm onClose={this.handleClick} transaction={transaction} {...other} />;
    }

    return <Item onClick={this.handleClick} {...transaction} />;
  }
}

export default TransactionListItem;
