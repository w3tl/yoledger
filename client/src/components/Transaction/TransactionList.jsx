import React from 'react';
import PropTypes from 'prop-types';
import TransactionListItem from './TransactionListItem';
import { transactionPropType } from './propTypes';

function TransactionList({ transactions, ...other }) {
  return (
    <ul>
      {transactions.map(trans => (
        <TransactionListItem key={trans.id} transaction={trans} {...other} />))}
    </ul>
  );
}

TransactionList.propTypes = {
  transactions: PropTypes.arrayOf(transactionPropType),
};

TransactionList.defaultProps = {
  transactions: [],
};

export default TransactionList;
