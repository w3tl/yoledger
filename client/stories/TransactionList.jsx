import React from 'react';
import { storiesOf } from '@storybook/react';

import TransactionList from '../src/components/Transaction/TransactionList';

const transactions = [{
  id: '1',
  amount: 10,
  source: { name: 'Cash' },
  destination: { name: 'Food' },
  date: new Date('2018-01-01').toString(),
}, {
  id: '2',
  amount: 10,
  source: { name: 'Cash' },
  destination: { name: 'Bank Card' },
  date: new Date('2018-04-01').toString(),
}, {
  id: '3',
  amount: 10,
  source: { name: 'Bank Card' },
  destination: { name: 'Food' },
  date: new Date('2018-01-10').toString(),
}];

storiesOf('Transaction/List', module)
  .add('with data', () => <TransactionList transactions={transactions} />)
  .add('empty list', () => <TransactionList />);
