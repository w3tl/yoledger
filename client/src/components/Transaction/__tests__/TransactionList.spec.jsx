import React from 'react';
import { shallow } from 'enzyme';
import TransactionList from '../TransactionList';
import TransactionListItem from '../TransactionListItem';
import transactions from '../mockData';

describe('TransactionList', () => {
  it('should render correctly', () => {
    const output = shallow(
      <TransactionList transactions={transactions} />,
    );
    expect(output).toMatchSnapshot();
    expect(output.find(TransactionListItem)).toHaveLength(transactions.length);
  });
});
