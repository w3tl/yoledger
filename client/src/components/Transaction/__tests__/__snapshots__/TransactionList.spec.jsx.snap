// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList should render correctly 1`] = `
<ul>
  <TransactionListItem
    key="1"
    transaction={
      Object {
        "__typename": "Transaction",
        "amount": 10,
        "date": "2018-01-01T00:00:00.000Z",
        "destination": Object {
          "__typename": "Account",
          "id": "21",
          "name": "Food",
        },
        "id": "1",
        "source": Object {
          "__typename": "Account",
          "id": "2",
          "name": "Cash",
        },
      }
    }
  />
  <TransactionListItem
    key="2"
    transaction={
      Object {
        "__typename": "Transaction",
        "amount": 10,
        "date": "2018-04-01T00:00:00.000Z",
        "destination": Object {
          "__typename": "Account",
          "id": "1",
          "name": "Bank Card",
        },
        "id": "2",
        "source": Object {
          "__typename": "Account",
          "id": "2",
          "name": "Cash",
        },
      }
    }
  />
  <TransactionListItem
    key="3"
    transaction={
      Object {
        "__typename": "Transaction",
        "amount": 10,
        "date": "2018-01-06T00:00:00.000Z",
        "destination": Object {
          "__typename": "Account",
          "id": "21",
          "name": "Food",
        },
        "id": "3",
        "source": Object {
          "__typename": "Account",
          "id": "1",
          "name": "Bank Card",
        },
      }
    }
  />
</ul>
`;
